token-lending/program/src/state/reserve.rs

```diff
@@ -74,306 +89,260 @@ impl Reserve {
-        }
-
-        // Special handling for small, closeable obligations
-        let max_closeable_amount = obligation.max_closeable_amount()?;
-        let close_amount = liquidity_amount.min(max_closeable_amount);
-        if close_amount > 0 {
-            return Ok(LiquidateResult {
-                withdraw_amount: obligation.deposited_collateral_tokens,
-                repay_amount: close_amount,
-                settle_amount: obligation.borrowed_liquidity_wads,
-            });
+    /// Update borrow rate and accrue interest
+    pub fn accrue_interest(&mut self, current_slot: Slot) -> ProgramResult {
+        let slots_elapsed = self.last_update.slots_elapsed(current_slot)?;
+        if slots_elapsed > 0 {
+            let current_borrow_rate = self.current_borrow_rate()?;
+            self.liquidity
+                .compound_interest(current_borrow_rate, slots_elapsed)?;
         }
+        Ok(())
+    }
 
-        // Calculate the amount of liquidity that will be repaid
-        let max_liquidation_amount = obligation.max_liquidation_amount()?;
-        let repay_amount = liquidity_amount.min(max_liquidation_amount);
-        let decimal_repay_amount = Decimal::from(repay_amount);
-
-        // Calculate the amount of collateral that will be received
-        let withdraw_amount = {
-            let receive_liquidity_amount =
-                token_converter.convert(decimal_repay_amount, liquidity_token_mint)?;
-            let collateral_amount = collateral_exchange_rate
-                .decimal_liquidity_to_collateral(receive_liquidity_amount)?;
-            let bonus_rate = Rate::from_percent(collateral_reserve_config.liquidation_bonus);
-            let bonus_amount = collateral_amount.try_mul(bonus_rate)?;
-            let withdraw_amount = collateral_amount.try_add(bonus_amount)?;
-            let withdraw_amount =
-                withdraw_amount.min(obligation.deposited_collateral_tokens.into());
-            if repay_amount == max_liquidation_amount {
-                withdraw_amount.try_ceil_u64()?
-            } else {
-                withdraw_amount.try_floor_u64()?
+    /// Borrow liquidity up to a maximum market value
+    pub fn borrow_liquidity(
+        &self,
+        liquidity_amount: u64,
+        max_borrow_value: Decimal,
+    ) -> Result<BorrowLiquidityResult, ProgramError> {
+        let decimals = 10u64
+            .checked_pow(self.liquidity.mint_decimals as u32)
+            .ok_or(LendingError::MathOverflow)?;
+        if liquidity_amount == u64::max_value() {
+            let borrow_amount = max_borrow_value
+                .try_mul(decimals)?
+                .try_div(self.liquidity.market_price)?
+                .min(self.liquidity.available_amount.into());
+            let (origination_fee, host_fee) = self
+                .config
+                .fees
+                .calculate_borrow_fees(borrow_amount, FeeCalculation::Inclusive)?;
```

Comment:
In this design, fees do contribute to the borrowed total. Per discussion with @bartosz-lipinski's, `borrow_amount` in the instruction reflects the amount the borrower wants to borrow, i.e. the tokens they will receive. When `borrow_amount` is `u64::MAX` as it is here, instead the maximum borrow amount is calculated, fees are calculated inclusive of the total and subtracted to get the `receive_amount`.

The total (receive amount + fees) come out of the reserve's available liquidity, and the obligation's `borrowed_amount_wads` is set to this total, so the borrower will pay back the fees (eventually). The upsides of this approach are that it's simple, and the fees are paid to the reserve that the liquidity is borrowed from.

However a downside of this approach is that it means that the borrower pays nothing initially, but the fee receiver is paid right away. If the fee receiver is, in effect, the liquidity pool, this seems fine. However, if this is leads to undesirable outcomes, we could have fees come out of one of the collaterals deposited.

The borrower would need to provide collateral accounts when borrowing, and the calculation is more complicated. This also makes less sense to me, since it means the collateral deposit reserve's fee receiver is going to receive fees when liquidity is borrowed from another reserve, but I could be mistaken about the purpose of the fees.

Reply:
Ah, I missed that the fees were added to the outstanding balance of the loan. That makes sense, thanks for the explanation!

Reply:
I'm being nit picky about naming, but I think what threw me off is the name of `borrow_amount`. It's the amount that's borrowed from the reserve, sure, but it really represents the amount that will be added to the loan balance. I'm not too savvy but financial terms, but is there something more clear we could call this?

Reply:
I'm not sure what else to call it! The `borrow_amount` is both the total amount that will be borrowed from the reserve and the total amount that will be added to the loan balance. It's exactly what equals or is added to `borrowed_amount_wads` in both cases. The internally used `receive_amount` is the amount the borrower will receive.

Reply:
Cool, that seems fine then. This is probably something that could be solved by documentation